sstp_protocol/src/charset.rs

```diff
@@ -0,0 +1,113 @@
+use std::fmt;
+
+/// Error that can occur when converting Charset from string.
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(
+        "invalid charset `{0}`, the charset must be ASCII, Shift_JIS, ISO-2022-JP, EUC-JP or UTF-8"
+    )]
+    Invalid(String),
+}
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, PartialEq, PartialOrd, Copy, Clone, Eq, Ord, Hash)]
+enum Inner {
+    Ascii,
+    ShiftJis,
+    Iso2022Jp,
+    EucJp,
+    Utf8,
+}
+
+/// Charset is the character set and encoding of strings in SSTP headers.
+///
+/// ```rust
+/// # use sstp_protocol::Charset;
+/// #
+/// # let charset = Charset::ASCII;
+/// match charset {
+///     Charset::ASCII => assert_eq!(charset.to_string(), "ASCII"),
+///     Charset::SHIFT_JIS => assert_eq!(charset.to_string(), "Shift_JIS"),
+///     Charset::ISO2022JP => assert_eq!(charset.to_string(), "ISO-2022-JP"),
+///     Charset::EUC_JP => assert_eq!(charset.to_string(), "EUC-JP"),
+///     Charset::UTF8 => assert_eq!(charset.to_string(), "UTF-8"),
+/// }
+/// ```
+#[derive(Debug, PartialEq, PartialOrd, Copy, Clone, Eq, Ord, Hash)]
+pub struct Charset(Inner);
```

Comment:
This was created in reference to [http](https://github.com/hyperium/http/blob/master/src/version.rs).
But I don't see the need to add an enum to struct as you suggest.
I would change the code to enum with strum.

Reply:
I understood the change to use strum and enum.
There is a problem with using enum.

`HeaderName` requires `Other(String)` with user-specified header.
```rust
pub enum HeaderName {
  ...
  Other(String)
}
```

The string that `Other` contains is a restricted string based on RFC7230.
```rust
HeaderName::Other("foo")
```

However, since the type is a string, the user is free to set it. This is a problem.

How about using struct only for `HeaderName`?
```rust
let request = builder
    .notify(Version::Sstp10)
    .header(HeaderName::SENDER, "Sakura")
    .header(HeaderName::EVENT, "OnMusicPlay")
    .header(HeaderName::REFERENCE0, "The Legend of the Original Takagi Bou")
    .header(HeaderName::REFERENCE1, "Muscle Girl Band")
    .charset(Charset::ShiftJis)
    .build()
    .unwrap();
```

This mixes the camel case and snake case in the builder and looks bad.

Do you have a better way to use enum?
If there is no way to do this, I think it is best to use struct and enum.

Reply:
> Point 1: HeaderName requires validation for the Other(String) variant
> It is worth noting that from_str would still be accessible if the enum is exposed publicly, which may require documentation instructing users to use from_unvalidated_str.

I think this is the worst way to code a type safe.
Do not trade type safety for boilerplate.

> Point 2: Camel-case Headers

I'm sorry. I should have said Pascal Case and Screaming Snake Case.

```rust
let request = builder
    .notify(Version::Sstp10)
    .header(HeaderName::SENDER, "Sakura")
    .header(HeaderName::EVENT, "OnMusicPlay")
    .header(HeaderName::REFERENCE0, "The Legend of the Original Takagi Bou")
    .header(HeaderName::REFERENCE1, "Muscle Girl Band")
    .charset(Charset::ShiftJis)
    .build()
    .unwrap();
```

My problem is that the builder has a `Version` and `Charset` of Pascal Case, but the `HeaderName` is Screaming Snake Case.

This is the correct naming convention for Rust.
However, the code is very difficult to write because it is not consistent across `Version`, `Charset`, and `HeaderName`.

To me, it is more important to have consistent input for the builder's argument in either the Pascal case or the Snake case than it is to turn off the boilerplate.

Reply:
> Please demonstrate the specific issue with type safety that you're referring to.

It is possible to call `HeaderName::Other(String)` to create a `HeaderName` with a string that does not conform to RFC7230.
The interface should be exposed so that `HeaderName` types always have only RFC7230-compliant strings.

> If this is such a high priority that it supersedes Rust's naming conventions, then just turn off the warnings by putting this at the top of each relevant file:

No. The Rust naming convention should prevail.
I am saying that we should create types `Version`, `Charset`, and `HeaderName` so that we can pass consistent arguments in the builder.

Reply:
> It seems you may have missed the final two lines of that message, please read it again more carefully.

Thank you! I finally understood.
You are saying to use strum in the inner pattern.
That would certainly be safer and reduce SLOCs.

Reply:
> Could you please push code (either to this branch or a separate branch) that demonstrates the issue you're encountering which prevents you from implementing the builder in the way that you want?

```rust
let request = builder
    .notify(Version::SSTP_10)
    .header(HeaderName::SENDER, "Sakura")
    .header(HeaderName::EVENT, "OnMusicPlay")
    .header(HeaderName::REFERENCE0, "The Legend of the Original Takagi Bou")
    .header(HeaderName::REFERENCE1, "Muscle Girl Band")
    .charset(Charset::SHIFT_JIS)
    .build()
    .unwrap();
```

or

```rust
let request = builder
    .notify(Version::Sstp10)
    .header(HeaderName::Sender, "Sakura")
    .header(HeaderName::Event, "OnMusicPlay")
    .header(HeaderName::Reference0, "The Legend of the Original Takagi Bou")
    .header(HeaderName::Reference1, "Muscle Girl Band")
    .charset(Charset::ShiftJis)
    .build()
    .unwrap();
```

I believe that the type should be created so that it can be either of these.

```rust
let request = builder
    .notify(Version::Sstp10)
    .header(HeaderName::SENDER, "Sakura")
    .header(HeaderName::EVENT, "OnMusicPlay")
    .header(HeaderName::REFERENCE0, "The Legend of the Original Takagi Bou")
    .header(HeaderName::REFERENCE1, "Muscle Girl Band")
    .charset(Charset::ShiftJis)
    .build()
    .unwrap();
```

I don't want to mix Pascal case and Snake case in the values I pass as arguments like this.


So I can solve this by making everything an inner pattern with strum.